nni/algorithms/hpo/tpe_tuner.py

```diff
@@ -0,0 +1,374 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+"""
+Tree-structured Parzen Estimator (TPE) tuner for hyper-parameter optimization.
+
+Paper: https://proceedings.neurips.cc/paper/2011/file/86e8f7ab32cfd12577bc2619bc635690-Paper.pdf
+Official code: https://github.com/hyperopt/hyperopt/blob/master/hyperopt/tpe.py
+
+This is a slightly modified re-implementation of the algorithm.
+"""
+
+__all__ = ['TpeTuner', 'TpeArguments', 'suggest', 'suggest_parameter']
+
+from collections import defaultdict
+import logging
+import math
+from typing import NamedTuple, Optional, Union
+
+import numpy as np
+from scipy.special import erf  # pylint: disable=no-name-in-module
+
+from nni.tuner import Tuner
+from nni.common.hpo_utils import OptimizeMode, format_search_space, deformat_parameters
+from . import random_tuner
+
+_logger = logging.getLogger('nni.tuner.tpe')
+
+## Public API part ##
+
+class TpeArguments(NamedTuple):
+    """
+    These are the hyper-parameters of TPE algorithm itself.
+    To avoid confusing with trials' hyper-parameters, they are called "arguments" in this code.
+
+    Parameters
+    ==========
+    constant_liar_type: 'best' | 'worst' | 'mean' | None (default: 'best')
+        TPE algorithm itself does not support parallel tuning.
+        This parameter specifies how to optimize for trial_concurrency > 1.
+
+        None (or "null" in YAML) means do not optimize. This is the default behavior in legacy version.
+
+        How each liar works is explained in paper's section 6.1.
+        In general "best" suit for small trial number and "worst" suit for large trial number.
```

Comment:
better to also show this sentence in doc

Reply:
The doc has not been updated yet.
I had put a work item in iteration plan.